packages/blocks/src/api/templates.js

```diff
@@ -57,7 +58,20 @@ export function synchronizeBlocksWithTemplate( blocks = [], template ) {
 	return map( template, ( [ name, attributes, innerBlocksTemplate ], index ) => {
 		const block = blocks[ index ];
 
-		if ( block && block.name === name ) {
+		// Return the same block,
+		if (
+			// if block exists,
+			block &&
+			// and the name of the block is same,
+			block.name === name &&
+			// and
+			(
+				// either the template is given without any block attributes (this means the template doesn't care about the content).
+				! attributes ||
+				// or the attributes are same as what is given in the template.
+				isShallowEqual( attributes, block.attributes )
```

Comment:
I think the issue is that some templates do define attributes as initial value and this would break that behavior.

Reply:
If the template is defining the attributes then they will be overridden in the newly created block below. So should be fine. Isn't it?

Maybe I'm missing the point. Can you give me an example of it?

Reply:
Say I have this template:

```js
const template = [
  [ 'core/paragraph', { 'content', 'some initial random content' } ],
]
```

If I'm understanding the change here, everytime the post using this template as locked will be loaded, it will reset the content of the paragraph to "some initial random content". Am I wrong?

Reply:
You’re right.

Are you raising concern about backward compatibility?

If the template is changed midway, then the existing content will go away. But isn’t it that case as well, when the order of blocks is changed in the template? That’s what [this](https://github.com/WordPress/gutenberg/blob/master/packages/blocks/src/api/test/templates.js#L152) unit test is suggesting.

My understanding is, if the template is used with locking, then it’s implicit that whatever the template is forcing, it will be applied, irrespective of the existing content.

I’ve attached a screencast of what I’m trying to achieve. That may help in understanding my use case.

![2019-06-27 12 21 56](https://user-images.githubusercontent.com/2236554/60229157-4d5fe980-98d8-11e9-979d-88701723bd1d.gif)

Reply:
Yes, that's my concern. This PR is basically changing the meaning of the `templateLock="all"`. It was always meant to be something used to lock the block used but not their content even if you provide initial content.

Reply:
Okay. Then what do you have to say about this? 👇🏼

> But isn’t it that case as well, when the order of blocks is changed in the template? That’s what this unit test is suggesting.

Reply:
yes, locking locks the order as well. It's how it was designed. For example you want to render a list of blocks in a product template with a given order, lock everything and only allow the user to edit the content.

Reply:
Sorry, I'm coming back to this after some time. But just not able to get my head around the current behaviour.

You said that we should not change any existing content of any blocks irrespective of `templateLock` is applied or not.

But that's already happening if the order of blocks within a template is changed.

E.g.,

```
const templates = [
	[
		[
			'core/heading',
			{ content: name }
		],
		[
			'core/quote',
			{ value: company }
		],
		[
			'core/paragraph',
			{ content: `The balance is - ${ balance }` }
		]
	],
	[
		[
			'core/quote',
			{ value: company }
		],
		[
			'core/paragraph',
			{ content: `The balance is - ${ balance }` }
		],
		[
			'core/heading',
			{ content: name }
		]
	],
];

const template = templates[ Math.floor( Math.random() * Math.floor( 2 ) ) ];

const ALLOWED_BLOCKS = [ 'core/paragraph' ];

return (
	<InnerBlocks
		template={ template }
		templateLock="all"
		allowedBlocks={ ALLOWED_BLOCKS }
	/>
);
```

Above snippet is changing the blocks along with the content in v6.1.1

Isn't that contradictory? If not, then my understanding of allowing the template to override the content in any case is valid.